rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -0,0 +1,356 @@
+
+    @staticmethod
+    def _unfreeze_states(states: Deque[FrozenState]) -> List[State]:
+        return [
+            {key: dict(value) for key, value in dict(state).items()} for state in states
+        ]
+
+    def _create_states(
+        self, tracker: DialogueStateTracker, domain: Domain
+    ) -> List[State]:
+        """Create states: a list of dictionaries."""
+
+        states = tracker.past_states(domain)
+
+        return self._unfreeze_states(states)
+
+    def _featurize_states(
+        self,
+        trackers_as_states: List[List[State]],
+        interpreter: NaturalLanguageInterpreter,
+    ) -> List[List[Dict[Text, List["Features"]]]]:
+        return [
+            [
+                self.state_featurizer.encode_state(state, interpreter)
+                for state in tracker_states
+            ]
+            for tracker_states in trackers_as_states
+        ]
+
+    @staticmethod
+    def _convert_labels_to_ids(
+        trackers_as_actions: List[List[Text]], domain: Domain
+    ) -> np.ndarray:
+        # use numpy arrays since in `Features` there are array
+        return np.array(
+            [
+                np.array(
+                    [domain.index_for_action(action) for action in tracker_actions]
+                )
+                for tracker_actions in trackers_as_actions
+            ]
+        )
+
+    def training_states_and_actions(
+        self, trackers: List[DialogueStateTracker], domain: Domain
+    ) -> Tuple[List[List[State]], List[List[Text]]]:
+        """Transforms list of trackers to lists of states and actions."""
+
+        raise NotImplementedError(
+            "Featurizer must have the capacity to encode trackers to feature vectors"
+        )
+
+    def featurize_trackers(
+        self,
+        trackers: List[DialogueStateTracker],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> Tuple[List[List[Dict[Text, List["Features"]]]], np.ndarray]:
+        """
+        Featurize the training trackers.
+
+        Args:
+            trackers: list of training trackers
+            domain: the domain
+            interpreter: the interpreter
+
+        Returns:
+            - a dictionary of attribute (INTENT, TEXT, ACTION_NAME, ACTION_TEXT,
+              ENTITIES, SLOTS, ACTIVE_LOOP) to a list of features for all dialogue
+              turns in all training trackers
+            - the label ids (e.g. action ids) for every dialuge turn in all training
+              trackers
+        """
+        if self.state_featurizer is None:
+            raise ValueError(
+                f"Instance variable 'state_featurizer' is not set. "
+                f"During initialization set 'state_featurizer' to an instance of "
+                f"'{SingleStateFeaturizer.__class__.__name__}' class "
+                f"to get numerical features for trackers."
+            )
+
+        self.state_featurizer.prepare_from_domain(domain)
```

Comment:
is this rather a `re-initialize`?

Reply:
what do you mean?

Reply:
what `prepare_from_domain` does is reset and initialize the `featurizer`, isn't it?

Reply:
yes, because during creation of it, we don't know the domain

Reply:
but then it's rather `initialize` then `prepare`, right? Consider this non-blocking.

Reply:
I don't know, it's an old legacy name. But also I'd say that, the fact that it is re-init, is rather coincidence, there could be a featurizer that has more staff in init. This method is really domain related